### PR TITLE
unknown tokens in tokenizer encode input

### PR DESCRIPTION
**Pull request by floccinauc**
Added an option to control the behavior of modular tokenizer (and its operator FastTokenizer) when they encounter unknown tokens - ModularTokenizer.encode(), ModularTokenizer.encode_list() and FastTokenizer.__call__() have a parameter on_unknown:str='warn' that decides whether to raise a warning or an exception.
Also added ModularTokenizer.count_unknowns(encoding) that, given an encoding (or a list of ids), returns how many unknown tokens are in it.

### CODE REVIEW

File: fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py [hunk 1]
```diff
@@ -1046,6 +1075,13 @@ def encode_list(
                     f"both padding token and padding id are None, but padding length is {max_len}. It's possible that it was set for truncation alone."
                 )
 
+        # Check if we have any unknown tokens in our input
+        if self.count_unknowns(merged_encoding) > 0:
+            if on_unknown == "raise":
+                raise Exception("Encountered unknown tokens in the input")
+            elif on_unknown == "warn":
+                warn("Encountered unknown tokens in the input")
```
Comment (YoelShoshan) [inline fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py:1083]: please add another "else" and raise Exception if on_unknown got an unexpected value (and print what the user provided)
Additionally, consider adding 'ignore' option.

**Review by YoelShoshan — COMMENTED** [review]
Looks good!
few comments inline

File: fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py [hunk 2]
```diff
@@ -1046,6 +1075,13 @@ def encode_list(
                     f"both padding token and padding id are None, but padding length is {max_len}. It's possible that it was set for truncation alone."
                 )
 
+        # Check if we have any unknown tokens in our input
+        if self.count_unknowns(merged_encoding) > 0:
```
Comment (SagiPolaczek) [inline fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py:1079]: Do you think it'll be time consuming to count in each encoding?
If so maybe we should consider an option `on_unkown="ignore" / "skip" / "pass" , etc`, and make skipping the counting optional

Just a thought

File: fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py [hunk 3]
```diff
@@ -1046,6 +1075,13 @@ def encode_list(
...
+            if on_unknown == "raise":
+                raise Exception("Encountered unknown tokens in the input")
```
Comment (SagiPolaczek) [inline fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py:1081]: I suggest to prepare a more detailed error / warning msg. 
The number of `<UNK>` or the input itself would be helpful

**Review by SagiPolaczek — COMMENTED** [review]
Thanks Vadim!!
Added some minors too 😸

File: fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py (hunk 1 above)
Comment (floccinauc) [reply fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py]: done

File: fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py [hunk 4]
```diff
@@ -938,6 +942,29 @@ def get_expected_max_len(
...
+            unk_token (Optional[str], optional): Unknown token string (usually "<UNK>"). If None, locates the token in special_tokens.special_tokens.
+                Defaults to None.
+
+        Returns:
+            int: count of unknown tokens in the encoding
+        """
+        if isinstance(encoding, list):
+            ids = encoding
+        elif isinstance(encoding, Encoding):
+            ids = encoding.ids
```
[the remark this answers is not in the document]
Comment (floccinauc) [reply fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py:962]: Done

File: fusedrug/data/tokenizer/modulartokenizer/test_modular_tokenizer.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    test_tokenizer,
+)
+from fusedrug.data.tokenizer.modulartokenizer.modular_tokenizer import TypedInput
+
+# add cell type:
+config_name = "tokenizer_config_with_celltype"
+
+
+@hydra.main(config_path="./configs", config_name="tokenizer_config", version_base=None)
+def main(cfg: DictConfig) -> None:
```
[the remark this answers is not in the document]
Comment (floccinauc) [reply fusedrug/data/tokenizer/modulartokenizer/test_modular_tokenizer.py:15]: It's not intended as a unit test, just a manual capabilities test, that we can base future unit test on.

File: fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py (hunk 2 above)
Comment (floccinauc) [reply fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py:1079]: It's just a single pass over the ID list - should be negligible to the stuff going on during encode call.

File: fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py (hunk 3 above)
Comment (floccinauc) [reply fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py]: The trouble is that our input is a list of TypedInput, and to fully parse it just for the exception would be a bit too much for such a small function. As a compromise, I changed this to only output the input string of first element of the list. This should identify the input sufficiently in most cases.

File: fusedrug/data/tokenizer/modulartokenizer/test_modular_tokenizer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (SagiPolaczek) [reply fusedrug/data/tokenizer/modulartokenizer/test_modular_tokenizer.py:15]: It's really good to have this manual test, but attaching it to the Jenkins' run will allow to maintain it and make sure it won't break. Anyway the distance to a unit-test is quite small and can be done in the future :)

**Review by SagiPolaczek — APPROVED** [review]
LGTM, thanks!

One general comment regarding the modular tokenizer is that we should aim to add the tests to the Jenkins run, to maintain functionally and making sure it won't break

**Pull request merged**